### PR TITLE
Refactoring of TimeConstraints

### PR DESCRIPTION
**Pull request by rkolpakov**

### CODE REVIEW

File: test/unit/scripts/upgrade/TimeConstraints.t.sol [hunk 1]
```diff
@@ -74,57 +76,108 @@ contract TimeConstraintsTest is Test {
...
 
-        timeConstraints.checkExecuteWithinDayTime(Durations.from(18 hours), Durations.from(12 hours));
+        timeConstraints.checkTimeWithinDayTime(Durations.from(18 hours), Durations.from(12 hours));
+    }
+
+    // checkTimeWithinDayTimeAndEmit
+
+    function test_checkTimeWithinDayTimeAndEmit_HappyPath() external {
+        vm.warp(BASE_TIMESTAMP + 15 hours);
+        vm.expectEmit(true, true, true, true);
```
Comment (Psirex) [inline test/unit/scripts/upgrade/TimeConstraints.t.sol:104]: Overload of `vm.expectEmit()` without args by default checks all topics

File: test/unit/scripts/upgrade/TimeConstraints.t.sol [hunk 2]
```diff
@@ -74,57 +76,108 @@ contract TimeConstraintsTest is Test {
...
-    function testFuzz_checkExecuteAfterTimestamp_NotReached(uint32 secondsBefore) public {
+    function testFuzz_checkTimeAfterTimestamp_NotReached(uint32 secondsBefore) public {
         vm.assume(secondsBefore < BASE_TIMESTAMP && secondsBefore > 0);
 
         vm.warp(BASE_TIMESTAMP - secondsBefore);
         vm.expectRevert(
             abi.encodeWithSelector(TimeConstraints.TimestampNotReached.selector, Timestamps.from(BASE_TIMESTAMP))
         );
-        timeConstraints.checkExecuteAfterTimestamp(Timestamps.from(BASE_TIMESTAMP));
+        timeConstraints.checkTimeAfterTimestamp(Timestamps.from(BASE_TIMESTAMP));
```
Comment (Psirex) [inline test/unit/scripts/upgrade/TimeConstraints.t.sol:138]: `vm.expectRevert()` should be used with external call, to work correctly in the new forge versions.

File: test/unit/scripts/upgrade/TimeConstraints.t.sol [hunk 3]
```diff
@@ -74,57 +76,108 @@ contract TimeConstraintsTest is Test {
...
             )
         );
 
-        timeConstraints.checkExecuteWithinDayTime(Durations.from(18 hours), Durations.from(12 hours));
+        timeConstraints.checkTimeWithinDayTime(Durations.from(18 hours), Durations.from(12 hours));
+    }
+
+    // checkTimeWithinDayTimeAndEmit
+
+    function test_checkTimeWithinDayTimeAndEmit_HappyPath() external {
```
Comment (Psirex) [inline test/unit/scripts/upgrade/TimeConstraints.t.sol:102]: Add also fuzz tests for the `checkTimeWithinDayTimeAndEmit()` method

File: test/unit/scripts/upgrade/TimeConstraints.t.sol [hunk 4]
```diff
@@ -74,57 +76,108 @@ contract TimeConstraintsTest is Test {
...
         vm.expectRevert(
             abi.encodeWithSelector(TimeConstraints.TimestampExceeded.selector, Timestamps.from(BASE_TIMESTAMP))
         );
-        timeConstraints.checkExecuteBeforeTimestamp(Timestamps.from(BASE_TIMESTAMP));
+        timeConstraints.checkTimeBeforeTimestamp(Timestamps.from(BASE_TIMESTAMP));
+    }
+
+    // checkTimeBeforeTimestampAndEmit
+
+    function testFuzz_checkTimeBeforeTimestampAndEmit_HappyPath() public {
```
Comment (Psirex) [inline test/unit/scripts/upgrade/TimeConstraints.t.sol:172]: This test doesn't use fuzzing

File: scripts/upgrade/TimeConstraints.sol [hunk 5]
```diff
@@ -47,22 +55,44 @@ contract TimeConstraints {
...
+    /// @param endDayTime The end time of the allowed range in seconds since midnight (UTC).
+    function checkTimeWithinDayTimeAndEmit(Duration startDayTime, Duration endDayTime) external {
+        checkTimeWithinDayTime(startDayTime, endDayTime);
+        emit PerformedWithinDayTime(startDayTime, endDayTime);
+    }
+
     /// @notice Checks that the transaction can only be executed after a specific timestamp.
     /// @param timestamp The Unix timestamp after which the function can be executed.
-    function checkExecuteAfterTimestamp(Timestamp timestamp) external view {
+    function checkTimeAfterTimestamp(Timestamp timestamp) public view {
```
Comment (Psirex) [inline scripts/upgrade/TimeConstraints.sol:68]: This method currently checks whether the current time is greater than or equal to the timestamp, which doesn’t match the method name. Let’s update the condition in the if statement to align with the method’s intended behavior.

Comment (Psirex) [reply scripts/upgrade/TimeConstraints.sol:68]: Same in the `checkTimeBeforeTimestamp()` method

File: test/unit/scripts/upgrade/TimeConstraints.t.sol [hunk 6]
```diff
@@ -74,57 +76,108 @@ contract TimeConstraintsTest is Test {
...
-        timeConstraints.checkExecuteBeforeTimestamp(Timestamps.from(BASE_TIMESTAMP));
+        timeConstraints.checkTimeBeforeTimestamp(Timestamps.from(BASE_TIMESTAMP));
 
         vm.warp(BASE_TIMESTAMP - 1 hours);
-        timeConstraints.checkExecuteBeforeTimestamp(Timestamps.from(BASE_TIMESTAMP));
+        timeConstraints.checkTimeBeforeTimestamp(Timestamps.from(BASE_TIMESTAMP));
     }
 
-    function testFuzz_checkExecuteBeforeTimestamp_Exceed(uint32 secondsAfter) public {
+    function testFuzz_checkTimeBeforeTimestamp_Exceed(uint32 secondsAfter) public {
```
Comment (Psirex) [inline test/unit/scripts/upgrade/TimeConstraints.t.sol:160]: Such a fuzzy implementation limits the range of cases being tested. Instead of adding seconds to a timestamp, the test can take two parameters — chainTime and timestamp, both of type Timestamp. Assuming that chainTime < timestamp, perform the check accordingly.

File: scripts/upgrade/TimeConstraints.sol [hunk 7]
```diff
@@ -33,7 +41,7 @@ contract TimeConstraints {
     /// @notice Checks that the transaction can only be executed within a specific time range during the day.
     /// @param startDayTime The start time of the allowed range in seconds since midnight (UTC).
     /// @param endDayTime The end time of the allowed range in seconds since midnight (UTC).
-    function checkExecuteWithinDayTime(Duration startDayTime, Duration endDayTime) external view {
+    function checkTimeWithinDayTime(Duration startDayTime, Duration endDayTime) public view {
```
[the remark this answers is not in the document]
Comment (Psirex) [reply scripts/upgrade/TimeConstraints.sol:44]: Worth also adding dev comment, that range check is inclusive

**Review by Psirex — CHANGES_REQUESTED** [review]

File: scripts/upgrade/TimeConstraints.sol [hunk 8]
```diff
@@ -8,6 +8,14 @@ import {Timestamps, Timestamp} from "contracts/types/Timestamp.sol";
 /// @title Time Constraints Contract
 /// @notice Provides functionality to restrict execution of functions based on time constraints.
 contract TimeConstraints {
+    // ---
+    // Events
+    // ---
+
+    event PerformedWithinDayTime(Duration startDayTime, Duration endDayTime);
```
[the remark this answers is not in the document]
Comment (rkolpakov) [reply scripts/upgrade/TimeConstraints.sol:15]: Maybe just `WithinDayTimeChecked()`, `BeforeTimestampChecked()` and `AfterTimestampChecked()`?

For the second two there are no specific time we want to check, it may be several days or weeks.

File: scripts/upgrade/TimeConstraints.sol [hunk 9]
```diff
@@ -30,39 +36,69 @@ contract TimeConstraints {
...
+        checkTimeAfterTimestamp(timestamp);
+        emit TimeAfterTimestampChecked(timestamp);
+    }
+
+    /// @notice Checks that the current timestamp is before the given specific timestamp.
     /// @param timestamp The Unix timestamp before which the function can be executed.
-    function checkExecuteBeforeTimestamp(Timestamp timestamp) external view {
-        if (Timestamps.now() > timestamp) {
+    function checkTimeBeforeTimestamp(Timestamp timestamp) public view {
+        if (Timestamps.now() >= timestamp) {
```
Comment (Psirex) [inline scripts/upgrade/TimeConstraints.sol:90]: `TimestampPassed(timestamp)` for error name, to match the `checkAfterTimestamp()` semantic

File: scripts/upgrade/TimeConstraints.sol [hunk 10]
```diff
@@ -30,39 +36,69 @@ contract TimeConstraints {
...
-        if (Timestamps.now() < timestamp) {
+    function checkTimeAfterTimestamp(Timestamp timestamp) public view {
+        if (Timestamps.now() <= timestamp) {
             revert TimestampNotReached(timestamp);
         }
     }
 
-    /// @notice Checks that the transaction can only be executed before a specific timestamp.
+    /// @notice Checks that the current timestamp is after the given specific timestamp and emits an event.
+    /// @param timestamp The Unix timestamp after which the function can be executed.
```
Comment (Psirex) [inline scripts/upgrade/TimeConstraints.sol:81]: Still used "functions execution" semantics in natspec

File: test/unit/scripts/upgrade/TimeConstraints.t.sol [hunk 11]
```diff
@@ -3,7 +3,7 @@ pragma solidity 0.8.26;
 
 import {Test} from "forge-std/Test.sol";
 import {Durations, Duration} from "contracts/types/Duration.sol";
-import {Timestamps, Timestamp} from "contracts/types/Timestamp.sol";
+import {Timestamps, Timestamp, toSeconds} from "contracts/types/Timestamp.sol";
```
Comment (Psirex) [inline test/unit/scripts/upgrade/TimeConstraints.t.sol:6]: `Timestamps` and `toSeconds` is not used

File: test/unit/scripts/upgrade/TimeConstraints.t.sol [hunk 12]
```diff
@@ -3,7 +3,7 @@ pragma solidity 0.8.26;
...
 import {TimeConstraints} from "scripts/upgrade/TimeConstraints.sol";
 
 contract TimeConstraintsTest is Test {
```
Comment (Psirex) [inline test/unit/scripts/upgrade/TimeConstraints.t.sol:9]: `test_DAY_DURATION()` - Function state mutability can be restricted to view

File: test/unit/scripts/upgrade/TimeConstraints.t.sol [hunk 13]
```diff
@@ -34,97 +34,182 @@ contract TimeConstraintsTest is Test {
...
-        timeConstraints.checkExecuteWithinDayTime(Durations.from(12 hours), Durations.from(18 hours));
-
-        vm.warp(BASE_TIMESTAMP + 12 hours);
-        timeConstraints.checkExecuteWithinDayTime(Durations.from(12 hours), Durations.from(18 hours));
-
-        vm.warp(BASE_TIMESTAMP + 18 hours);
-        timeConstraints.checkExecuteWithinDayTime(Durations.from(12 hours), Durations.from(18 hours));
+    // checkTimeWithinDayTime
+
+    function testFuzz_checkTimeWithinDayTime_HappyPath_Regular(
```
Comment (Psirex) [inline test/unit/scripts/upgrade/TimeConstraints.t.sol:39]: This test narrows the chain time to the range [0, 24 hours]. Update the test to allow the full `uint40` range for the chain time, for example:
```solidity
    function testFuzz_checkTimeWithinDayTime_HappyPath_Regular(uint32 from, uint32 to, uint40 timeToWarp) external {
        from = from % 24 hours;
        to = to % 24 hours;
        uint256 warpDayTime = timeToWarp % 24 hours;

        vm.assume(from <= to && warpDayTime >= from && warpDayTime <= to);

        vm.warp(timeToWarp);
        timeConstraints.checkTimeWithinDayTime(Durations.from(from), Durations.from(to));
    }
```

Comment (Psirex) [reply test/unit/scripts/upgrade/TimeConstraints.t.sol:39]: Use such an approach accross all tests instead of adding time to `BASE_TIMESTAMP`

File: test/unit/scripts/upgrade/TimeConstraints.t.sol [hunk 14]
```diff
@@ -34,97 +34,182 @@ contract TimeConstraintsTest is Test {
...
-            )
+            abi.encodeWithSelector(TimeConstraints.DayTimeOutOfRange.selector, currentTime, startTime, endTime)
         );
-
-        timeConstraints.checkExecuteWithinDayTime(Durations.from(12 hours), Durations.from(18 hours));
+        this.external__checkTimeWithinDayTime(startTime, endTime);
     }
 
-    function test_checkExecuteWithinDayTime_TimeOverflow() external {
+    function test_checkTimeWithinDayTime_TimeOverflow() external {
```
Comment (Psirex) [inline test/unit/scripts/upgrade/TimeConstraints.t.sol:101]: This test also may use fuzzing

File: test/unit/scripts/upgrade/TimeConstraints.t.sol [hunk 15]
```diff
@@ -34,97 +34,182 @@ contract TimeConstraintsTest is Test {
...
-                TimeConstraints.InvalidDayTimeRange.selector, Durations.from(18 hours), Durations.from(12 hours)
-            )
-        );
+    // checkTimeWithinDayTimeAndEmit
+
+    function testFuzz_checkTimeWithinDayTimeAndEmit_HappyPath_Regular(
+        Duration from,
+        Duration to,
+        Duration timeToWarp
+    ) external {
```
Comment (Psirex) [inline test/unit/scripts/upgrade/TimeConstraints.t.sol:115]: More reliably, call `testFuzz_checkTimeWithinDayTime_HappyPath(from, to, timeToWarp)` together with checking that an event is emitted.

Comment (Psirex) [reply test/unit/scripts/upgrade/TimeConstraints.t.sol:115]: Same for the other tests which check emitted events

Comment (Psirex) [reply test/unit/scripts/upgrade/TimeConstraints.t.sol:115]: It would also be nice to have tests for the edge cases, for example, `from == to`, `from = 24 hours - 1, to = 0`, `to = 0, from = 24 hours - 1`, etc.

File: test/unit/scripts/upgrade/TimeConstraints.t.sol [hunk 16]
```diff
@@ -34,97 +34,182 @@ contract TimeConstraintsTest is Test {
...
+
+    function testFuzz_checkTimeAfterTimestamp_HappyPath(Timestamp timestampToWarp, Timestamp timestampToCheck) public {
+        vm.assume(timestampToWarp > timestampToCheck);
+        vm.warp(timestampToWarp.toSeconds());
+        timeConstraints.checkTimeAfterTimestamp(timestampToCheck);
     }
 
-    function testFuzz_checkExecuteAfterTimestamp_NotReached(uint32 secondsBefore) public {
-        vm.assume(secondsBefore < BASE_TIMESTAMP && secondsBefore > 0);
+    function testFuzz_checkTimeAfterTimestamp_NotReached(
```
Comment (Psirex) [inline test/unit/scripts/upgrade/TimeConstraints.t.sol:146]: Preserve naming convention for tests: `testFuzz_checkTimeAfterTimestamp_RevertOn_TimestampNotPassed()`

File: test/unit/scripts/upgrade/TimeConstraints.t.sol [hunk 17]
```diff
@@ -34,97 +34,182 @@ contract TimeConstraintsTest is Test {
...
-        timeConstraints.checkExecuteAfterTimestamp(Timestamps.from(BASE_TIMESTAMP));
+        this.external__checkTimeAfterTimestamp(timestampToCheck);
     }
 
-    function test_checkExecuteBeforeTimestamp_HappyPath() public {
-        vm.warp(BASE_TIMESTAMP);
-        timeConstraints.checkExecuteBeforeTimestamp(Timestamps.from(BASE_TIMESTAMP));
+    // checkTimeAfterTimestampAndEmit
+
+    function testFuzz_checkTimeAfterTimestampAndEmit_HappyPath(
```
Comment (Psirex) [inline test/unit/scripts/upgrade/TimeConstraints.t.sol:161]: Add the same set of negative tests for the "Emit" methods as for the "regular" check methods.

**Review by Psirex — CHANGES_REQUESTED** [review]

**Review by Psirex — COMMENTED** [review]
LGTM 👍🏻 I left a couple of minor comments — nothing critical, but worth addressing.

**Review by Psirex — APPROVED** [review]
LGTM👍🏻
